Feature: Mainly for updates of Personalization

ChenYiLins opened the pull request:
## Feature: Mainly for updates of WallpaperPickerPage

### Describe

Refer to TimePage designed by @Jay-o-Way. The function is basically consistent with the original Page.

### Screenshot

#### Take an overall view

![Take an overall view](https://github.com/user-attachments/assets/f4dc8efb-6399-433f-8cff-ab1c87697bce)

#### Picture

![Picture](https://github.com/user-attachments/assets/1c6fa914-61ce-40ef-81ae-492993c2cd6f)

#### Picture mm

![Picturemm](https://github.com/user-attachments/assets/8b2d6be7-9623-48c5-91c5-5edac49507a6)

#### Solid color

![Solid color](https://github.com/user-attachments/assets/e40e7232-0702-4bab-aa37-17f656cec99e)

Jay-o-Way commented:
Hm, I have an idea now. I think **Mode** and the settings immediately below that, could also be in one **Expander**. But low priority.

Jay-o-Way commented:
Honestly, I don't think the ComboButton is appropriate here. User can set things for _both_ themes, it's not one-or-the-other.

Armin2208 commented:
> Honestly, I don't think the ComboButton is appropriate here. User can set things for _both_ themes, it's not one-or-the-other.

Maybe you can realize this with a tab interface in this case.

Jay-o-Way commented on AutoDarkModeApp/ViewModels/WallpaperPickerViewModel.cs:
```diff
@@ -36,21 +84,153 @@ public WallpaperPickerViewModel(IErrorService errorService)
+            switch (_builder.Config.WallpaperSwitch.Component.TypeLight)
+            {
+                case WallpaperType.Global:
+                    IsPictureType = true;
```
What do these variables do? Why do they exist, when they _seem_ to do the same as the previous variable?

Jay-o-Way commented on AutoDarkModeApp/ViewModels/WallpaperPickerViewModel.cs:
```diff
@@ -36,21 +84,153 @@ public WallpaperPickerViewModel(IErrorService errorService)
+        IsPictureMMType = false;
+        IsSolidColorType = false;
+        IsSpotlightType = false;
+        ImageSettingsCardVisibility = Visibility.Collapsed;
```
Looks like you created a variable, but why not set `.Visibility` directly? Or use the Enum in the x:Bind?

ChenYiLins replied:
You are right. It's time to replace it to Enum.

Jay-o-Way commented on AutoDarkModeApp/ViewModels/WallpaperPickerViewModel.cs:
```diff
@@ -76,4 +256,147 @@ partial void OnIsWallpaperSwitchEnabledChanged(bool value)
+
+    partial void OnIsPictureTypeChanged(bool value)
+    {
+        if (SelectIndexWallpaperMode == 0 && value == true)
```
Is it an idea to use an Enum, instead of zero, one..?

Jay-o-Way commented on AutoDarkModeApp/ViewModels/WallpaperPickerViewModel.cs:
```diff
@@ -76,4 +256,147 @@ partial void OnIsWallpaperSwitchEnabledChanged(bool value)
+        }
+    }
+
+    partial void OnIsSpotlightTypeChanged(bool value)
```
These are very repetitive. There must be a simpler way.

ChenYiLins commented:
> Honestly, I don't think the ComboButton is appropriate here. User can set things for _both_ themes, it's not one-or-the-other.

Actually, only one theme setting will take effect at a time. In the original code, the option of this ComboBox initializes with the current theme of the computer, which is very intuitive.

But your point of view is actually very correct. In the original Pick an accent color Page, it used the method you mentioned.

If we use ComboBox, it may be ambiguous to users, because users will not know for the first time that different types can be adopted under different themes.

![image](https://github.com/user-attachments/assets/aaab45ed-2a35-4340-b5d1-2e075b193e71)

Jay-o-Way reviewed (commented):
Quickly from my phone

Jay-o-Way commented:
> Maybe you can realize this with a tab interface in this case.

Or maybe a [Selector bar](https://learn.microsoft.com/en-us/windows/apps/design/controls/selector-bar) ? Let's give this UX a good night sleep.

ChenYiLins commented on AutoDarkModeApp/Views/WallpaperPickerPage.xaml:
```diff
@@ -9,54 +9,160 @@
+                                <Button
+                                    Grid.Row="1"
+                                    MinWidth="120"
+                                    HorizontalAlignment="Right"
```
Sad  ( ; ´д｀)ゞ This is did happen.

![image](https://github.com/user-attachments/assets/dc23fed2-887d-4ae6-8d5b-51c68c7a0c06)

Spiritreader commented:
The new wallpaper page looks nice! 
Do you know if it's still possible to provide an image preview like with the old UI?

A couple of notes:

> In the original code, the option of this ComboBox initializes with the current theme of the computer, which is very intuitive.

This is really important and mandatory. Otherwise it becomes really confusing when trying to configure it (believe me, I experienced this a lot of times before we implemented this)

ChenYiLins commented:
## Code quality: Improve WallpaperPickerPage

### Describe

- Change bool variable to enum.
- Add picture preview.
- Add Grid column definition to avoid overlapping controls.
- Increase Border display color.
- On the initialization page, set ThemeMode to follow the current system theme.
- Add converter: EnumToIndexConverter, FlagsToVisibilityConverter.

> This is really important and mandatory. Otherwise it becomes really confusing when trying to configure it (believe me, I experienced this a lot of times before we implemented this)

It's right. We use ComboBox is the better choice. Because in the configuration file, not all settings correspond to Light or Dark Mode.

```yaml
  Component:
    Position: Fill
    Monitors:
    - Id: '\\?\DISPLAY#SKG2725#7&e9a60b&0&UID256#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7}'
```

### ScreenShots

#### Picture preview

![image](https://github.com/user-attachments/assets/b496603e-f8e6-4b94-8998-409aa772af91)

#### Avoid overlapping controls

![image](https://github.com/user-attachments/assets/6ac31123-5d99-4def-89b1-fe1960a4101c)

#### Colorful border

![image](https://github.com/user-attachments/assets/5c2895dc-86f6-4158-85ac-eaa42e090743)

ChenYiLins commented:
## Code quality: Improve performance

### Describe

- Fix missing enum.
- Prevent the page from being loaded twice because of the change of the value.
- Optimized the code for loading the monitors.
- The method of OnNavigatedFrom is added to View, which makes it possible to remove HandleConfigUpdate when the page leaves. (It's very import! This will cause very serious performance problems! ) (I feel more and more wrong, because it takes a long time to load the page, so I found this problem. (っ °Д °;)っ)

### Supplement

In fact, loading monitors in ViewModel will also cause some performance consumption. This is the test result:
```
Load monitors
Time:104
Time:61
Time:58
Time:55
Time:79

Do not load monitors
Time:22
Time:7
Time:3
Time:26
Time:3
```

Maybe we should separate this method and avoid using x:bind, but GlobalWallpaperPath depends on SelectMonitor (MonitorSettings), which in turn depends on Monitors. My mood: （；´д｀）ゞ

ChenYiLins commented on AutoDarkModeApp/Views/WallpaperPickerPage.xaml:
```diff
@@ -7,56 +7,209 @@
+                                    <ColumnDefinition Width="*" />
+                                </Grid.ColumnDefinitions>
+
+                                <Border
```
Because not every block needs a dividing line, such as Spotlight.

ChenYiLins commented on AutoDarkModeApp/ViewModels/WallpaperPickerViewModel.cs:
```diff
@@ -1,54 +1,216 @@
+        });
     }
 
+    public void OnNaOnNavigatedFrom(NavigationEventArgs e) => StateUpdateHandler.OnConfigUpdate -= HandleConfigUpdate;
```
At the same time, there is another interesting point here, about INavigationAware? If you use it, I don't know why, when you use the back button of the navigation bar, it doesn't trigger OnNavigatedFrom.

Jay-o-Way replied:
The buttons (Pane and Back) in the Navigation control are hidden. Their input is taken over by the two buttons in the XAML TitleBar. Both have a function to connect to the Navigation.

ChenYiLins replied:
You are right, that's why we can't use INavigaionAware, I will delete it. And more, I renamed it now.

Jay-o-Way commented on AutoDarkModeApp/Views/WallpaperPickerPage.xaml:
```diff
@@ -7,56 +7,209 @@
+                                <RadioButton Content="{helpers:ResourceString Name=WallpaperComboBoxItemPicture}" />
+                                <RadioButton Content="{helpers:ResourceString Name=WallpaperComboBoxItemPictureMM}" />
+                                <RadioButton Content="{helpers:ResourceString Name=WallpaperComboBoxItemSolidColor}" />
+                                <RadioButton Content="{helpers:ResourceString Name=ThemePickerTheme11Spotlight}" />
```
The full string should be "Windows Spotlight". Extra idea: add a Hyperlink to `ms-settings:personalization-background` ?

ChenYiLins replied:
Thanks for your idea. Although not all dividing lines can be merged, because multi-monitors and wallpaper filling methods are only suitable for multi-monitors mode. In the test, I can't seem to use `NavigateUri`, which can't trigger correctly. So I ues `Click`.

ChenYiLins replied:
I found [this](https://github.com/microsoft/microsoft-ui-xaml/issues/10428).

ChenYiLins commented:
## Code quality: Add WindowsSpotlight hyperlink

### Describe

- Merage dividing lines.
- Add a HyperlinkButton to navigate to Windows settings.

### Screenshots

![image](https://github.com/user-attachments/assets/ac095c13-6cd0-45b8-8500-9a70968aa7fa)

Jay-o-Way commented on AutoDarkModeApp/Helpers/EnumToIndexConverte.cs:
Filename: missing an "R"

ChenYiLins replied:
OK! Fixed.

Jay-o-Way commented on AutoDarkModeApp/Views/WallpaperPickerPage.xaml:
```diff
@@ -142,7 +146,7 @@
                                 <Grid.ColumnDefinitions>
                                     <ColumnDefinition Width="*" />
-                                    <ColumnDefinition Width="32" />
+                                    <ColumnDefinition Width="36" />
```
Why not use **Auto**?

Jay-o-Way commented on AutoDarkModeApp/Views/WallpaperPickerPage.xaml:
```diff
@@ -18,12 +19,15 @@
-                <controls:SettingsCard Header="{helpers:ResourceString Name=WallpaperComboBoxMode}" Visibility="{x:Bind ViewModel.IsWallpaperSwitchEnabled, Mode=OneWay}">
+                <controls:SettingsCard
+                    Header="{helpers:ResourceString Name=WallpaperComboBoxMode}"
+                    HeaderIcon="{ui:FontIcon Glyph=&#xE790;}"
```
How do you feel about icon E793? I think this fits for the dark/light context.

ChenYiLins replied:
To be honest, I referred to the official applications of Win11, such as calculator and notepad.

![image](https://github.com/user-attachments/assets/01f99019-391a-4a4f-8a1f-e4e6071590df)

And I try to ues this icon in other place.

![image](https://github.com/user-attachments/assets/ed363418-80e5-42de-a9e9-fc9429d792f1)

Jay-o-Way replied:
Good reason!

ChenYiLins commented:
## Code quality: Renameed

### Describe

- Buttons should convey their meaning through more detailed words, rather than icons.
- Rename some names to make them more uniform.

### Screenshots

![image](https://github.com/user-attachments/assets/9ca6e61b-ceb2-4bc3-98d5-0057db8f9ed8)

ChenYiLins commented:
## Feature: Mainly for updates of ColorizationPage

### Describe

Add ColorizationPage.

### Screenshots

![image](https://github.com/user-attachments/assets/b598ac00-14b0-4ee5-adf8-d1136daff259)

![gif](https://github.com/user-attachments/assets/028ac412-1b3d-442f-90a5-bee7b63d8bc6)

ChenYiLins commented:
> Idea: if it were me, I would say [ Color ... Select color ]

@Jay-o-Way I found a example.

![image](https://github.com/user-attachments/assets/0afb99b2-aa31-43d8-87ea-6c5af31ed82f)

Spiritreader commented:
> ## Code quality: Improve performance
> ### Describe
> 
>     * Fix missing enum.
> 
>     * Prevent the page from being loaded twice because of the change of the value.
> 
>     * Optimized the code for loading the monitors.
> 
>     * The method of OnNavigatedFrom is added to View, which makes it possible to remove HandleConfigUpdate when the page leaves. (It's very import! This will cause very serious performance problems! ) (I feel more and more wrong, because it takes a long time to load the page, so I found this problem. (っ °Д °;)っ)
> 
> 
> ### Supplement
> 
> In fact, loading monitors in ViewModel will also cause some performance consumption. This is the test result:
> 
> ```
> Load monitors
> Time:104
> Time:61
> Time:58
> Time:55
> Time:79
> 
> Do not load monitors
> Time:22
> Time:7
> Time:3
> Time:26
> Time:3
> ```
> 
> Maybe we should separate this method and avoid using x:bind, but GlobalWallpaperPath depends on SelectMonitor (MonitorSettings), which in turn depends on Monitors. My mood: （；´д｀）ゞ

Yeah, getting monitor information from the windows API is unexplicably slow for some reason, which is why there is this caching mechanism. Previously, loading the monitor settings was done asynchronously on page open. Then the page opened quickly  and the monitor strings were loaded in deferred.
Is that no longer an option with WinUi3?

Afaik there's no way to speed this operation up, so at some point getting the name string is necessary.

ChenYiLins commented:
> Yeah, getting monitor information from the windows API is unexplicably slow for some reason, which is why there is this [caching mechanism](https://github.com/ChenYiLins/Windows-Auto-Night-Mode/blob/79e51e4295c6b74bff2461317ee3c4887ac35895/AutoDarkModeLib/ComponentSettings/Base/WallpaperSwitchSettings.cs#L76). Previously, loading the monitor settings was done asynchronously on page open. Then the page opened quickly and the monitor strings were loaded in deferred. Is that no longer an option with WinUi3?
> 
> Afaik there's no way to speed this operation up, so at some point getting the name string is necessary.

Actually, maybe I'm a little picky. IIn WinUI, we can completely rely on loading asynchronously, but we need to add something to View. Fortunately, the current processing method will not cause too much performance burden.

tooomm commented:
> Idea: if it were me, I would say [ Color ... Select color ]

![image](https://github.com/user-attachments/assets/4e92eef7-fffd-4c11-a100-be119d110c96)

I also feel the wording of the description and the button is not ideal.
`Accent color                  [ ] [ Select color ]` or `[ Color picker ]`

Spiritreader commented:
https://github.com/user-attachments/assets/693a35d1-d5f1-4693-83e8-07436c90f31e

Changing the monitor selection does not update the wallpaper path or preview

ChenYiLins commented:
## Feature: Add the missing function

### Describe

- Switching monitors options will now match the correct wallpaper.
- The unconnected monitors can be removed normally.
- Changed the method of loading the monitors. Due to asynchronous execution, there is almost no additional performance overhead for loading pages.
- Changed the display mode of TeachingTip, and now it will automatically close if it loses focus.

### Screenshots

![image](https://github.com/user-attachments/assets/e252d94c-26ea-4e1b-9bd1-b245c6450895)

![image](https://github.com/user-attachments/assets/31d73114-50ec-4e03-a890-db62e005ac52)
